mx_fetchjdk.py

```diff
@@ -0,0 +1,197 @@
+    full_jdk_path = distribution.get_full_jdk_path(jdk_path)
+    jdk_url = mx_urlrewrites.rewriteurl(distribution.get_url())
+    jdk_url_sha = jdk_url + ".sha1"
+    jdk_archive = distribution.get_archive()
+
+    if not exists(full_jdk_path):
+        archive_location = join(jdk_path, jdk_archive)
+        mx._opts.no_download_progress = args["quiet"]
+        try:
+            sha1_hash = mx._hashFromUrl(jdk_url_sha)
+        except:
+            mx.abort("Error fetching sha1 hash, check your download location")
+        if not exists(archive_location) or not args["keep-archive"]:
+            mx.download_file_with_sha1(jdk_artifact, archive_location, [jdk_url], sha1_hash, archive_location + '.sha1', resolve=True, mustExist=True, sources=False)
+        untar = mx.TarExtractor(archive_location)
+
+        if not args["quiet"]:
+            print("Installing...")
+
+        with untar._open() as tar_file:
+            curr_full_jdk_path = untar._getnames(tar_file)[0]
+
+        try:
+            untar.extract(jdk_path)
+        except:
+            os.remove(join(jdk_path, jdk_archive))
+            shutil.rmtree(curr_full_jdk_path)
+            mx.abort("Error parsing archive. Please try again")
+        if not args["keep-archive"]:
+            os.remove(join(jdk_path, jdk_archive))
+        os.rename(join(jdk_path, curr_full_jdk_path), full_jdk_path)
+
+    elif not args["quiet"]:
+        mx.warn("Requested JDK was already present")
+
+    if mx.is_darwin() and exists(join(full_jdk_path, 'Contents', 'Home')):
+        if args["strip-contents-home"]:
+            tmp_full_jdk_path = full_jdk_path + ".tmp"
+            shutil.move(full_jdk_path, tmp_full_jdk_path)
+            shutil.move(join(tmp_full_jdk_path, 'Contents', 'Home'), full_jdk_path)
+            shutil.rmtree(tmp_full_jdk_path)
+        else:
+            full_jdk_path = join(full_jdk_path, 'Contents', 'Home')
+
+    if "alias" in args:
+        alias_full_path = join(jdk_path, args["alias"])
+        if exists(alias_full_path) or islink(alias_full_path):
+            if isdir(alias_full_path) and not islink(alias_full_path):
+                shutil.rmtree(alias_full_path)
+            else:
+                os.remove(alias_full_path)
+
+        if not (mx.is_windows() or mx.is_cygwin()):
+            os.symlink(full_jdk_path, alias_full_path)
+        else:
+            copytree(full_jdk_path, alias_full_path, symlinks=True) # fallback for windows
+        full_jdk_path = alias_full_path
+
+    if not args["quiet"]:
+        print("Run the following to set JAVA_HOME in your shell:")
+
+    print('export JAVA_HOME={}'.format(abspath(full_jdk_path)))
+
+
+def _parse_fetchjdk_settings(args):
+    settings = {}
+    settings["quiet"] = False
+    settings["keep-archive"] = False
+    settings["jdk-path"] = abspath(join('bin', 'jdks'))
+
+    jdk_paths = find_system_jdks()
+    if len(jdk_paths) > 0:
+        settings["jdk-path"] = next(iter(jdk_paths))
```

Comment:
This makes the default installation location:
a) non-deterministic (`find_system_jdks()` returns a set)
b) the path of `JAVA_HOME` for an installed JDK where as the `--to` argument should be the base directory into which the JDK is installed. That is, `/Library/Java/JavaVirtualMachines` on macOS, `/usr/lib/jvm` on Linux, `/usr/jdk/instances` on Solaris and `C:\Program Files\Java` on Windows.

Reply:
Fixed